### DISCUSSION TITLE
List comprehension-like workflow?

### DISCUSSION THREAD

**Discussion by mihirmanna** [sharkdp/numbat#654] [Q&A] [answered]
I have lists of the masses, diameters, and heights of various aluminum cylinders and I want to calculate their respective densities. How can I achieve that using Numbat? This is what I've been trying so far.
```
let masses           = [85.2g,   264.3g,  438.1g,  441.7g,  441.9g]
let heights          = [0.99cm,  3.015cm, 4.99cm,  5cm,     4.99cm]
let diameters        = [6.325cm, 6.345cm, 6.350cm, 6.345cm, 6.350cm]

fn calculate_density(mass: Mass, height: Length, diameter: Length) -> MassDensity = mass / (pi*(diameter/2)^2 * height)

let rho = calculate_density(masses, heights, diameters)  # Would like to output a list of densities
```
But the type checker won't let me pass lists into the function. Is there a way to get around this? I'm pretty new to Numbat so I appreciate any input! Thanks 😁

**Comment by Goju-Ryu** [comment, marked as answer]
You can't easily do this with three lists, but if you make one list of [structs](https://numbat.dev/doc/structs.html) each containing one mass, height and diameter, then you can use the [map](https://numbat.dev/doc/list-functions-lists.html#map) function to apply a function on each element. 

If you need I can provide a simple example implementation of this later when I have a little more time.

**Reply by mihirmanna** [reply]
Thanks! I've implemented something like this for now:
```
struct Cylinder {
    mass:     Mass,
    height:   Length,
    diameter: Length,
}

let cylinders = [
    Cylinder { mass: 85.2g,  height: 0.99cm,  diameter: 6.325cm },
    Cylinder { mass: 264.3g, height: 3.015cm, diameter: 6.345cm },
    Cylinder { mass: 438.1g, height: 4.99cm,  diameter: 6.350cm },
    Cylinder { mass: 441.7g, height: 5cm,     diameter: 6.345cm },
    Cylinder { mass: 441.9g, height: 4.99cm,  diameter: 6.350cm }
]

fn density(cylinder: Cylinder) -> MassDensity = cylinder.mass / (pi*(cylinder.diameter/2)^2 * cylinder.height)

let densities = map(density, cylinders)
```

Which seems to do the trick.

**Reply by Goju-Ryu** [reply]
That is almost exactly the example I would have made if you had needed it 😄
